docs: Document entity token formatting with getFormattedTokens() method (#551)
mautic/developer-documentation-new: docs/plugin_extensions/emails.rst

.. vale off

Database-sourced tokens with BuilderTokenHelper
===============================================

.. vale on

> When your tokens represent database entities like Pages, Assets, or Forms, use the ``BuilderTokenHelper`` to query and format them. This helper handles permission checks, search filtering, and consistent label formatting.

Reviewer: Address the Vale warning: [Mautic.FeatureList] Is this referring to a Mautic feature? If so, use 'Landing Pages' instead of 'Pages'.

Getting the helper
------------------

Inject ``BuilderTokenHelperFactory`` and create a helper for your entity type:

.. code-block:: PHP

    <?php

    // plugins/HelloWorldBundle/EventListener/BuilderSubscriber.php

    declare(strict_types=1);

    namespace MauticPlugin\HelloWorldBundle\EventListener;

    use Mautic\CoreBundle\Helper\BuilderTokenHelperFactory;
    use Mautic\EmailBundle\EmailEvents;
    use Mautic\EmailBundle\Event\EmailBuilderEvent;
    use Symfony\Component\EventDispatcher\EventSubscriberInterface;

    final class BuilderSubscriber implements EventSubscriberInterface
    {
        public function __construct(
            private BuilderTokenHelperFactory $builderTokenHelperFactory,
        ) {
        }

        public static function getSubscribedEvents(): array
        {
            return [
                EmailEvents::EMAIL_ON_BUILD => ['onEmailBuild', 0],
            ];
        }

        public function onEmailBuild(EmailBuilderEvent $event): void
        {
            // Create a helper for the 'page' model
            $tokenHelper = $this->builderTokenHelperFactory->getBuilderTokenHelper('page');
        }
    }

> Formatting tokens with getFormattedTokens()
> -------------------------------------------

Reviewer: ```suggestion
.. vale off

Formatting tokens with getFormattedTokens()
-------------------------------------------

.. vale on
```

.. code-block:: PHP

    <?php

    use Mautic\CoreBundle\DTO\TokenFormatOptions;
    use Mautic\CoreBundle\Event\BuilderEvent;
    use Mautic\CoreBundle\Helper\BuilderTokenHelperFactory;

    public function onBuilderBuild(BuilderEvent $event): void
    {
        $tokenHelper = $this->builderTokenHelperFactory->getBuilderTokenHelper('page');
        $tokenFilter = $event->getTokenFilter();

        // Get tokens with formatted labels like "Page: My Landing Page (123)"
        $tokens = $tokenHelper->getFormattedTokens(
            'pagelink=(\d+)',                                              // Token regex pattern
            TokenFormatOptions::linkWithId('mautic.page.page', 'pagelink=(\d+)'), // Format options
            'label' === $tokenFilter['target'] ? $tokenFilter['filter'] : '',     // Search filter
            'title'                                                        // Label column
        );

        if ($tokens) {
            $event->addTokens($tokens);
        }
    }

Token format options
--------------------

The ``TokenFormatOptions`` class provides two factory methods for common label formats:

> **Simple prefix** - Creates labels like 'Form: My Form':

Reviewer: Address the Vale warnings:

- [Google.FirstPerson] Avoid first-person pronouns such as 'My'.
- [Google.Colons] ': M' should be in lowercase.

.. code-block:: PHP

    TokenFormatOptions::simplePrefix('mautic.form.form')
    // Result: "Form: Contact Form"

> **Link with ID** - Creates labels like 'Page: my-alias (123)' for linkable entities:

Reviewer: Address the Vale warnings:

- [Google.Parens] Use parentheses judiciously.
- [Google.FirstPerson] Avoid first-person pronouns such as 'my'.
- [Mautic.FeatureList] Is this referring to a Mautic feature? If so, use 'Landing Page' instead of 'Page'.

.. code-block:: PHP

    TokenFormatOptions::linkWithId('mautic.page.page', 'pagelink=(\d+)')
    // Result: "a:Page: my-landing-page (456)"

> The ``a:`` prefix indicates a link token. The second parameter is a regex pattern to extract the entity ID from the token.

Reviewer: ```suggestion
The ``a:`` prefix indicates a link token. The second parameter is a regular expression pattern to extract the entity ID from the token.
```

Token sorting
-------------

> Mautic automatically sorts tokens in the builder dropdown by category. The sort order is:

Reviewer: ```suggestion
.. vale off

Mautic automatically sorts tokens in the builder dropdown by category. The sort order is:

.. vale on
```

1. Contact fields
2. Company fields
3. Owner fields
4. Page links
5. Dynamic Web Content
6. Focus Items
7. Assets
8. This Page tokens (Landing Page builder only)
9. Email tokens

> 10. Other tokens

Reviewer: Follow our style guide.

```suggestion
.. vale off

#. Contact fields
#. Company fields
#. Owner fields
#. Page links
#. Dynamic Web Content
#. Focus Items
#. Assets
#. This Page tokens - Landing Page builder only
#. Email tokens
#. Other tokens

.. vale on
```

Within the Contact category, the following fields appear first: First Name, Last Name, and Title.

Within each category, tokens sort alphabetically by label.

Deprecated methods
------------------

> 

Author: PR #15855 deprecates `BuilderEvent::addTokensFromHelper()` and `BuilderEvent::getTokensFromHelper()` in favor of `BuilderTokenHelper::getFormattedTokens()`. Deprecation documented in `BuilderEvent.php` lines with `@deprecated` annotations.

Source: https://github.com/mautic/mautic/pull/15855

            TokenFormatOptions::linkWithId('mautic.page.token.pagelink', 'pagelink=(\d+)'),  // Format options
            'label' === $tokenFilter['target'] ? $tokenFilter['filter'] : '',                // Search filter
            'title'                                                                          // Label column
        );

        if ($tokens) {
            $event->addTokens($tokens);
        }
    }

Token format options
--------------------

The ``TokenFormatOptions`` class provides two factory methods for common label formats:

.. vale off

**Simple prefix** - Creates labels like 'Form: Contact Form':

.. vale on

.. code-block:: PHP

    TokenFormatOptions::simplePrefix('mautic.form.form')
    // Result: "Form: Contact Form"

.. vale off

**Link with ID** - Creates labels like 'Page link: my-landing-page (456)' for linkable entities:

.. vale on

.. code-block:: PHP

    TokenFormatOptions::linkWithId('mautic.page.token.pagelink', 'pagelink=(\d+)')
    // Result: "a:Page link: my-landing-page (456)"

The ``a:`` prefix indicates a link token. The second parameter is a regular expression pattern to extract the entity ID from the token.

> Token sorting

Reviewer 2: **[Low] Token-sorting list omits a category.** This list ("Contact, Company, Owner, Page link, Dynamic Content, Focus Item, Asset, Email") skips `CATEGORY_THIS_PAGE` (priority 80 in `TokenSorter.php`), which sits between Asset and Email. Low impact since the list is already framed as illustrative and defers to `TokenSorter` as authoritative — optional to fix, e.g. by adding "This page" into the sequence.

-------------

.. vale off

Mautic automatically groups tokens in the builder dropdown by type, such as Contact, Company, Owner, Page link, Dynamic Content, Focus Item, Asset, Email. Custom or unrecognized tokens fall into an 'Other' group.

.. vale on

Within the Contact group, First Name, Last Name, and Title appear first. Tokens then sort alphabetically by label within each group. The ``TokenSorter`` class constants in Mautic core define the authoritative order.

Deprecated methods
------------------

> .. deprecated:: 7.0

Reviewer 2: **[High] Wrong deprecation version.** This states `.. deprecated:: 7.0`, but `addTokensFromHelper()`/`getTokensFromHelper()` were deprecated by the linked core PR mautic/mautic#15855 (merge commit `118fd518a9`). I verified with `git describe --tags` on that commit that it merges *after* the `7.0.0`, `7.0.1`, `7.0.2`, `7.1.0`, `7.1.1`, `7.1.2`, and `7.1.3` tags — it first appears in `7.2.0-rc`, which also matches the core PR's own GitHub milestone (`7.2.0-rc`). The `BuilderEvent.php` docblock carries no version tag, so this looks like an incorrect inference. Please change to `.. deprecated:: 7.2`.

.. code-block:: PHP

    <?php

    use Mautic\CoreBundle\DTO\TokenFormatOptions;
    use Mautic\CoreBundle\Event\BuilderEvent;
    use Mautic\CoreBundle\Helper\BuilderTokenHelperFactory;

    public function onBuilderBuild(BuilderEvent $event): void
    {
        $tokenHelper = $this->builderTokenHelperFactory->getBuilderTokenHelper('page');
        $tokenFilter = $event->getTokenFilter();

        // Get tokens with formatted labels like "Page: My Landing Page (123)"
        $tokens = $tokenHelper->getFormattedTokens(
            'pagelink=(\d+)',                                              // Token regex pattern
            TokenFormatOptions::linkWithId('mautic.page.page', 'pagelink=(\d+)'), // Format options
            'label' === $tokenFilter['target'] ? $tokenFilter['filter'] : '',     // Search filter
            'title'                                                        // Label column
        );

        if ($tokens) {
            $event->addTokens($tokens);
        }
    }

Token format options
--------------------

The ``TokenFormatOptions`` class provides two factory methods for common label formats:

> **Simple prefix** - Creates labels like 'Form: contact form':

Reviewer: Address the Vale warnings:

- [Mautic.FeatureList] Is this referring to a Mautic feature? If so, use 'Form' instead of 'form'.
- [Mautic.FeatureList] Is this referring to a Mautic feature? If so, use 'Contact' instead of 'contact'.

.. code-block:: PHP

    TokenFormatOptions::simplePrefix('mautic.form.form')
    // Result: "Form: Contact Form"

**Link with ID** - Creates labels like 'Landing Page: alias 123' for linkable entities:

.. code-block:: PHP

    TokenFormatOptions::linkWithId('mautic.page.page', 'pagelink=(\d+)')
    // Result: "a:Page: my-landing-page (456)"

The ``a:`` prefix indicates a link token. The second parameter is a regular expression pattern to extract the entity ID from the token.

Token sorting
-------------

.. vale off

Mautic automatically sorts tokens in the builder dropdown by category. The sort order is:

.. vale on

.. vale off

#. Contact fields
#. Company fields
#. Owner fields
#. Page links
#. Dynamic Web Content
#. Focus Items
#. Assets
#. This Page tokens - Landing Page builder only
#. Email tokens
#. Other tokens

.. vale on

.. vale off

Within the Contact category, the following fields appear first: First Name, Last Name, and Title.

.. vale on

> Within each category, tokens sort alphabetically by label.

Reviewer: Move the vale on statement.

```suggestion
Within the Contact category, the following fields appear first: First Name, Last Name, and Title.

Within each category, tokens sort alphabetically by label.

.. vale on
```

    use Mautic\CoreBundle\DTO\TokenFormatOptions;
    use Mautic\CoreBundle\Event\BuilderEvent;
    use Mautic\CoreBundle\Helper\BuilderTokenHelperFactory;

    public function onBuilderBuild(BuilderEvent $event): void
    {
        $tokenHelper = $this->builderTokenHelperFactory->getBuilderTokenHelper('page');
        $tokenFilter = $event->getTokenFilter();

        // Get tokens with formatted labels like "Page: My Landing Page (123)"
        $tokens = $tokenHelper->getFormattedTokens(
            'pagelink=(\d+)',                                              // Token regex pattern
            TokenFormatOptions::linkWithId('mautic.page.page', 'pagelink=(\d+)'), // Format options
            'label' === $tokenFilter['target'] ? $tokenFilter['filter'] : '',     // Search filter
            'title'                                                        // Label column
        );

        if ($tokens) {
            $event->addTokens($tokens);
        }
    }

Token format options
--------------------

The ``TokenFormatOptions`` class provides two factory methods for common label formats:

.. vale off

**Simple prefix** - Creates labels like 'Form: Contact Form':

.. vale on

> .. code-block:: PHP

Reviewer 2: The description says 'Landing Page' but the example uses `mautic.page.page` which translates to 'Page' (see `app/bundles/PageBundle/Translations/en_US/messages.ini:21`). The result comment at line 183 correctly shows `a:Page: ...`. Either change the description to 'Page' or replace the example key with `mautic.page.token.pagelink` (which the core `PageBundle\BuilderSubscriber` uses and means 'Page link').

Token sorting
-------------

.. vale off

Mautic automatically sorts tokens in the builder dropdown by category. The sort order is:

.. vale on

.. vale off

#. Contact fields
#. Company fields
#. Owner fields
#. Page links
#. Dynamic Web Content
#. Focus Items
#. Assets
#. This Page tokens - Landing Page builder only
#. Email tokens
#. Other tokens

.. vale on

.. vale off

Within the Contact category, the following fields appear first: First Name, Last Name, and Title.

Within each category, tokens sort alphabetically by label.

.. vale on

Deprecated methods
------------------

.. deprecated:: 7.0

   The following ``BuilderEvent`` methods are deprecated and will be removed in a future version:

   - ``addTokensFromHelper()`` - Use ``BuilderTokenHelper::getFormattedTokens()`` and ``$event->addTokens()`` instead.
   - ``getTokensFromHelper()`` - Use ``BuilderTokenHelper::getFormattedTokens()`` instead.

**Migration example:**

.. code-block:: PHP

    // Before (deprecated):
    public function onEmailBuild(EmailBuilderEvent $event): void
    {
        $event->addTokensFromHelper($tokenHelper, $tokenRegex, 'label', 'alias');
    }

>     // After:

Reviewer 2: This call drops the `$valueColumn` parameter. The original code at line 235 passes `'alias'` as the value column, but `getFormattedTokens` defaults to `'id'` here, silently changing token keys (e.g. `{contactfield=email}` → `{contactfield=42}`). Add `, 'alias'` as the 5th argument to preserve the original behavior.

.. code-block:: PHP

    <?php

    use Mautic\CoreBundle\DTO\TokenFormatOptions;
    use Mautic\CoreBundle\Event\BuilderEvent;
    use Mautic\CoreBundle\Helper\BuilderTokenHelperFactory;

    public function onBuilderBuild(BuilderEvent $event): void
    {
        $tokenHelper = $this->builderTokenHelperFactory->getBuilderTokenHelper('page');
        $tokenFilter = $event->getTokenFilter();

        // Get tokens with formatted labels like "Page link: my-landing-page (123)"
        $tokens = $tokenHelper->getFormattedTokens(
            'pagelink=(\d+)',                                                // Token regex pattern
            TokenFormatOptions::linkWithId('mautic.page.token.pagelink', 'pagelink=(\d+)'),  // Format options
            'label' === $tokenFilter['target'] ? $tokenFilter['filter'] : '',                // Search filter
            'title'                                                                          // Label column
        );

        if ($tokens) {
            $event->addTokens($tokens);
        }
    }

Token format options
--------------------

The ``TokenFormatOptions`` class provides two factory methods for common label formats:

.. vale off

> **Simple prefix** - Creates labels like 'Form: Contact Form':

Reviewer: ```suggestion
**Simple prefix** - Creates labels like ``Form: Contact Form``:
```

.. vale on

.. code-block:: PHP

    TokenFormatOptions::simplePrefix('mautic.form.form')
    // Result: "Form: Contact Form"

.. vale off

> **Link with ID** - Creates labels like 'Page link: my-landing-page (456)' for linkable entities:

Reviewer: ```suggestion
**Link with ID** - Creates labels like ``Page link: my-landing-page (456)`` for linkable entities:
```

.. vale on

.. code-block:: PHP

    TokenFormatOptions::linkWithId('mautic.page.token.pagelink', 'pagelink=(\d+)')
    // Result: "a:Page link: my-landing-page (456)"

The ``a:`` prefix indicates a link token. The second parameter is a regular expression pattern to extract the entity ID from the token.

Token sorting
-------------

.. vale off

Mautic automatically groups tokens in the builder dropdown by type, such as Contact, Company, Owner, Page link, Dynamic Content, Focus Item, Asset, This page, Email. Custom or unrecognized tokens fall into an 'Other' group.

.. vale on

Within the Contact group, First Name, Last Name, and Title appear first. Tokens then sort alphabetically by label within each group. The ``TokenSorter`` class constants in Mautic core define the authoritative order.

Deprecated methods
------------------

.. deprecated:: 7.2

>    The following ``BuilderEvent`` methods are deprecated and will be removed in a future version:

Reviewer: There are a couple of things:

- Use active instead of passive voice
- Don't use 'will'
